qemu/tests/qemu_guest_agent.py

```diff
@@ -1326,6 +1326,105 @@ def gagent_check_with_fsfreeze(self, test, params, env):
         cmd_del_file = "%s %s" % (params["cmd_del"], tmp_file)
         session.cmd(cmd_del_file)
 
+    @error_context.context_aware
+    def gagent_check_with_selinux(self, test, params, env):
+        """
+        File operation via guest agent when selinux policy is in "Enforcing"
+         mode and "Permissive" mode.
+
+        Steps:
+        1) set selinux policy to "Enforcing" mode in guest
+        2) create and write content to temp file and non temp file
+        3) open the temp file with w+ mode and a+ mode
+        4) open the non temp file with w+ mode and a+ mode
+        5) set selinux policy to "Permissive" in guest
+        6) repeate step3-4
+        7) recovery the selinux policy
+        :param test: kvm test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment.
+        """
+        def result_check(guest_file, open_mode, selinux_mode):
+            """
+            Open guest file via guest agent with different open-mode
+            in different selinux policy mode.
+
+            1) if open temp file with append mode via guest agent,neither
+             selinux is enforce nor permissive,can open it
+            2) if selinux policy is permissive,guest agent can open
+             file
+            3) if selinux policy is enforcing,guest agent can't open
+             file except 1) condition
+
+            :param guest_file: guest file to open via guest agent
+            :param open_mode: open file mode,such as w+ and a+
+            :param selinux: selinux policy mode
+            """
+            error_context.context("Try to open the %s with %s mode via"
+                                  " guest agent in %s selinux policy." %
+                                  (guest_file, open_mode, selinux_mode),
+                                  logging.info)
+            temp_file_append = "/tmp" in guest_file and open_mode == "a+"
+            if selinux_mode == "Permissive" or temp_file_append:
+                ret_handle = self.gagent.guest_file_open(guest_file,
+                                                         mode=open_mode)
+                self.gagent.guest_file_write(ret_handle, content)
+                self.gagent.guest_file_flush(ret_handle)
+                self.gagent.guest_file_close(ret_handle)
+            else:
```

Comment:
'Permissive' and 'Enforcing' check actually does not share much, can be two separated functions, actually 'a+' and 'w+' can be included in one functions since you always check them together, like...
```
def result_check_permissive(guest_file):
    def check(temp_file_append, open_mode):
        ret_handle = self.gagent.guest_file_open(guest_file,
                                                         mode=open_mode)
        self.gagent.guest_file_write(ret_handle, content)
        self.gagent.guest_file_flush(ret_handle)
        self.gagent.guest_file_close(ret_handle)
    temp_file_append = "/tmp" in guest_file
    check(temp_file_append, 'a+')
    check(False, 'w+')

def result_check_enforcing(guest_file):
    def check(open_mode):
        try:
            self.gagent.guest_file_open(guest_file,
                                        mode=open_mode)
        except guest_agent.VAgentCmdError as detail:
            msg = r"failed to open file.*Permission denied"
            if not re.search(msg, str(detail)):
                test.fail("This is not the desired information: "
                          "('%s')" % str(detail))
        else:
            test.fail("When selinux policy is 'Enforcing', guest"
                      " agent should not open %s with %s mode." %
                      (guest_file, open_mode))
    check('a+')
    check('w+')
```

Reply:
Yes, put them into two functions is more clear.
But  there is an exception, when selinux is enforcing, open temp file with a+ mode, can success without 'permission denied' info.

Reply:
@xiagao The key point is they share not much implementation so it does not make sense to me to have them in one function, please reorganize it.

Reply:
@vivianQizhu  Updated.thx